axum/src/body/stream_body.rs

```diff
@@ -109,22 +107,178 @@ impl<S> fmt::Debug for StreamBody<S> {
 }
 
 impl<S> HttpBody for StreamBody<S>
+where
+    S: Stream,
+    S::Item: Into<Bytes>,
+{
+    type Data = Bytes;
+    type Error = Infallible;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        self.poll_next(cx).map(|option_chunk| option_chunk.map(Ok))
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        Poll::Ready(Ok(None))
+    }
+}
+
+impl<S> Stream for StreamBody<S>
+where
+    S: Stream,
+    S::Item: Into<Bytes>,
+{
+    type Item = Bytes;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project()
+            .stream
+            .get_pin_mut()
+            .poll_next(cx)
+            .map(|option_chunk| option_chunk.map(Into::into))
+    }
+}
+
+pin_project! {
+    /// An [`http_body::Body`] created from a fallible [`Stream`].
+    ///
+    /// The purpose of this type is to be used in responses. If you want to
+    /// extract the request body as a stream consider using
+    /// [`BodyStream`](crate::extract::BodyStream).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use axum::{
+    ///     Router,
+    ///     routing::get,
+    ///     body::{TryStreamBody, Bytes},
+    ///     response::{IntoResponse, Response},
+    /// };
+    /// use futures::stream::{self, Stream};
+    /// use std::io;
+    ///
+    /// async fn handler() -> Response {
+    ///     TryStreamBody::new(some_fallible_stream())
+    ///         .on_error(on_stream_error)
+    ///         .into_response()
+    /// }
+    ///
+    /// fn some_fallible_stream() -> impl Stream<Item = Result<Bytes, io::Error>> + Send + 'static {
+    ///     // ...
+    ///     # futures_util::stream::empty()
+    /// }
+    ///
+    /// fn on_stream_error(err: io::Error) {
+    ///     // ...
+    /// }
+    ///
+    /// let app = Router::new().route("/", get(handler));
+    /// # async {
+    /// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+    /// # };
+    /// ```
+    ///
+    /// [`Stream`]: futures_util::stream::Stream
+    pub struct TryStreamBody<S, T = DefaultOnError> {
+        #[pin]
+        stream: SyncWrapper<S>,
+        on_error: T,
+    }
```

Comment:
I don't love the name `TryStreamBody` here because it makes it sound the body produces errors, which it doesn't. Its the stream the might fail. Its really `(TryStream)Body` 😅

Reply:
Hm maybe. Although there is a convention to use "try" (`TryFuture`, `TryStream`, etc)